Unity/Champloo/Assets/Scripts/Player/Movement States/Movement Specials/OnSlide.cs

```diff
@@ -0,0 +1,160 @@
+            else if (controller.collisions.Right && velocity.x == 0 && velocity.y < 0)
+            {
+                velocity.x += velocity.y;
+                velocity.y = 0;
+            }
+            else if (controller.collisions.Left && velocity.x == 0 && velocity.y < 0)
+            {
+                velocity.x -= velocity.y;
+                velocity.y = 0;
+            }
+            else
+            {
+                //add in any potential -y velocity if there is any, otherwise just decay the speed
+                velocity.x += (goingRight ? (groundSlideDecay - velocity.y) : -(groundSlideDecay - velocity.y));
+                //velocity.y = 0;
+            }
+        }
+        return velocity;
+    }
+
+    public override MovementState DecideNextState(Vector3 velocity, Vector3 externalForces)
+    {
+        if (timingState == TimingState.DONE)
+        {
+            return GetSimulatedState();
+        }
+        //if InAir, or hit something from below
+        else if(GetSimulatedState() == GetComponent<InAir>() || (GetSimulatedState() == GetComponent<OnWall>() && controller.collisions.Above))
+        {
+            return GetComponent<InAir>();
+        }
+
+
+        return null;
+    }
+
+
+    public override void OnEnter(Vector3 inVelocity, Vector3 inExternalForces, out Vector3 outVelocity, out Vector3 outExternalForces)
+    {
+        //short out
+        if(GetSimulatedState() == GetComponent<InAir>())
+        {
+            timingState = TimingState.DONE;
+            outVelocity = inVelocity;
+            outExternalForces = inExternalForces;
+            return;
+        }
+        base.OnEnter(inVelocity, inExternalForces, out outVelocity, out outExternalForces);
+
+        goingUp = player.Velocity.y > 0;
+        goingRight = player.Velocity.x > 0;
+    }
+
+
+    protected override void OnStart()
+    {
+        //if player has somehow entered the air since the OnEnter method, short out again
+        if (GetSimulatedState() == GetComponent<InAir>())
+        {
+            timingState = TimingState.DONE;
+            return;
+        }
+
+        base.OnStart();
+
+        test = player.AimDirection;
+
+        goingUp = player.Velocity.y > 0;
+        goingRight = player.Velocity.x > 0;
+
+        //if a player isn't moving, get their aim direction/direction they are facing, otherwise you can only slide the way you are running
+        if (player.Velocity.x == 0) goingRight = player.AimDirection.x > 0;
+
+        
+        if (GetSimulatedState() == GetComponent<OnWall>())
+        {
+            appliedForce.y = goingUp ? wallSlideUpForce : -wallSlideDownForce;
+        }
+        else if (GetSimulatedState() == GetComponent<OnGround>())
+        {
+            appliedForce.y = 0;
+            appliedForce.x = goingRight ? groundSlideForce.x : -groundSlideForce.x;
+        }
+
+        player.ApplyForce(appliedForce);
```

Comment:
Doing all the force application in ApplyFriction is problematic as that method isn't tied to timing and is meant for general frame by frame changes. I would prefer to apply forces in OnStart so we don't have to get around our timing restrictions with checks in ApplyFriction